cmd/ncproxy/computeagent_cache.go

```diff
@@ -0,0 +1,71 @@
+package main
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+type computeAgentCache struct {
+	// lock for synchronizing read/write access to `cache`
+	rw sync.RWMutex
+	// mapping of container ID to shim compute agent ttrpc service
+	cache map[string]*computeAgentClient
+}
+
+func newComputeAgentCache() *computeAgentCache {
+	return &computeAgentCache{
+		cache: make(map[string]*computeAgentClient),
+	}
+}
+
+func (c *computeAgentCache) getAllAndClear() ([]*computeAgentClient, error) {
+	// set c.cache to nil first so that subsequent attempts to reads and writes
+	// return an error
+	c.rw.Lock()
+	cacheCopy := c.cache
+	c.cache = nil
+	c.rw.Unlock()
+
+	if cacheCopy == nil {
+		return nil, errors.New("cannot read from a nil cache")
+	}
+
+	results := []*computeAgentClient{}
+	for _, agent := range cacheCopy {
+		results = append(results, agent)
+	}
+	return results, nil
+
+}
+
+func (c *computeAgentCache) get(cid string) (*computeAgentClient, error) {
+	c.rw.RLock()
+	defer c.rw.RUnlock()
+	if c.cache == nil {
+		return nil, errors.New("cannot read from a nil cache")
```

Comment:
Can we define an error for this at the top and use this in place of the four `errors.New("cannot ___ from a nil cache")`

e.g. `return errInvalidCache` or similar.